contracts/marketplace/lib.rs

```diff
@@ -0,0 +1,299 @@
+        #[ink(topic)]
+        pub account_id: AccountId,
+        pub price: Balance,
+    }
+
+    #[ink(event)]
+    #[derive(Clone, Eq, PartialEq, Debug)]
+    pub struct Reset;
+
+    impl Error {
+        fn missing_role(role: Role) -> Self {
+            MissingRole(format!("{:?}", role))
+        }
+    }
+
+    impl From<ink_env::Error> for Error {
+        fn from(inner: ink_env::Error) -> Self {
+            Error::ContractCall(format!("{:?}", inner))
+        }
+    }
+
+    impl From<PSP22Error> for Error {
+        fn from(inner: PSP22Error) -> Self {
+            Error::PSP22TokenCall(format!("{:?}", inner))
+        }
+    }
+
+    impl AccessControlled for Marketplace {
+        type ContractError = Error;
+    }
+
+    impl Marketplace {
+        #[ink(constructor)]
+        pub fn new(
+            ticket_token: AccountId,
+            reward_token: AccountId,
+            starting_price: Balance,
+            min_price: Balance,
+            sale_multiplier: Balance,
+            auction_length: BlockNumber,
+        ) -> Self {
+            Self::ensure_role(Self::initializer())
+                .unwrap_or_else(|e| panic!("Failed to initialize the contract {:?}", e));
+
+            Marketplace {
+                ticket_token,
+                reward_token,
+                min_price,
+                sale_multiplier,
+                auction_length,
+                current_start_block: Self::env().block_number(),
+                total_proceeds: starting_price.saturating_div(sale_multiplier),
+                tickets_sold: 1,
+            }
+        }
+
+        #[ink(message)]
+        /// The length of each auction of a single ticket in blocks.
+        ///
+        /// The contract will decrease the price linearly from `average_price() * sale_multiplier()`
+        /// to `min_price()` over this period. The auction doesn't end after the period elapses -
+        /// the ticket remains available for purchase at `min_price()`.
+        pub fn auction_length(&self) -> BlockNumber {
+            self.auction_length
+        }
+
+        #[ink(message)]
+        /// The block at which the auction of the current ticket started.
+        pub fn current_start_block(&self) -> BlockNumber {
+            self.current_start_block
+        }
+
+        #[ink(message)]
+        /// The price the contract would charge when buying at the current block.
+        pub fn price(&self) -> Balance {
+            self.current_price()
+        }
+
+        #[ink(message)]
+        /// The average price over all sales the contract made.
+        pub fn average_price(&self) -> Balance {
+            self.total_proceeds.saturating_div(self.tickets_sold)
+        }
+
+        #[ink(message)]
+        /// The multiplier applied to the average price after each sale.
+        ///
+        /// The contract tracks the average price of all sold tickets and starts off each new
+        /// auction at a price that is this multiplier times the average.
```

Comment:
```suggestion
        /// auction at a price that is `self.average_price() * self.sale_multiplier()`
```

Personally, that's cleaner to me.